content/6_Advanced/XOR_Basis.mdx

```diff
@@ -27,6 +26,244 @@ frequency: 1
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define i23 long long
+
+vector<i23> basis;
+
+void add(i23 x) {
+	for (int i = 0; i < (int)basis.size();
+	     i++) {  // reduce x using the current basis vectors
+		x = min(x, x ^ basis[i]);
+	}
+
+	if (x != 0) { basis.push_back(x); }
+}
+
+int main() {
+	int n;
+	cin >> n;
+	vector<i23> arr(n);
+	for (auto &x : arr) {
+		cin >> x;
+		add(x);
+	}
+
+	i23 res = (1LL << (int)basis.size()) - n;
+	cout << res << '\n';
+
+	return 0;
+}
+```
+</CPPSection>
+</LanguageSection>
+## Example - Trees and XOR Queries Again
+<FocusProblem problem="xorqueries" />
+
+You are given a tree consisting of $n$ vertices. There is an integer written on each vertex; the $i$-th vertex has integer $$a_i$$ written on it. You have to process $q$ queries. The $i$-th query consists of three integers $$x_i$$, $$y_i$$ and $$k_i$$. For this query, you have to answer if it is possible to choose a set of vertices $v_1,v_2,…,v_m$ (possibly empty) such that:
+- every vertex $$v_j$$ is on the simple path between $$x_i$$ and $$y_i$$ (endpoints can be used as well);
+- $$a_{v_1} \oplus a_{v_2} \oplus \dots \oplus a_{v_m} = k_i$$, where $\oplus$ denotes the bitwise XOR operator.
+
+## Solution
+
+<!-- [Official Tutorial](https://codeforces.com/blog/entry/122951) -->
+
+To efficiently compute XOR bases on paths in a tree, we use a method involving tree rooting, **Lowest Common Ancestor (LCA)**, and properties of **XOR bases**. The process begins by rooting the tree and using LCA to split any path into two vertical paths. For each vertex $$v$$, we maintain a list of "interesting" vertices that significantly influence the XOR base when traversing from $$v$$ to the root. Due to the properties of XOR bases, these lists are small, with a maximum size of 20.
+
+The core idea is to build these lists for all vertices efficiently. For a vertex $$v$$, its list is derived from its parent's list. If $$v$$ adds a new element to the XOR base of its parent's list, $$v$$ is added to its list; otherwise, one element in the parent's list is replaced. This propagation ensures that the size of each list remains manageable and enables efficient construction of these lists in $\mathcal{O}(nB^2)$ time, where $$B$$ is the size of the XOR base, typically 20.
+
+When answering a query, the XOR base for any path is obtained by combining the lists from the two vertical paths derived from the LCA. This merging and computation can be done in $\mathcal{O}(B^2 + \log n)$ time per query, providing an efficient solution for the problem.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int N = 200005, K = 20;
```

Comment:
(optional) set N to only 20001 if necessary

Reply:
Done.